### ISSUE TITLE
Invalid results when using KNeighborsClassifier on a benchmark datasets with sklearnex 2021.3.0

### ISSUE THREAD

**Issue by ogrisel** [intel/scikit-learn-intelex#767]
**Describe the bug**

Using the following benchmark script. I observe 0 classification accuracy instead of the expected ~86%.

**To Reproduce**

- `bench_knn.py`:

```python
import numpy as np
from subprocess import check_output
import threadpoolctl
from pprint import pprint
from time import perf_counter
from pathlib import Path
import sklearn
from sklearn.neighbors import KNeighborsClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split


sklearn_path = Path(sklearn.__file__).parent
sklearn_version = sklearn.__version__
sklearn_git_branch = check_output(
    f'(cd {sklearn_path} && git symbolic-ref --short HEAD) || echo ""',
    shell=True,
).decode("utf-8").strip()
if sklearn_git_branch:
    sklearn_version = f"{sklearn_version} (git: {sklearn_git_branch})"
print(f"sklearn {sklearn_version} at {sklearn_path}")
pprint(threadpoolctl.threadpool_info())

X, y = make_classification(n_samples=int(2e6), n_features=100, random_state=0)
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    train_size=int(1e6),
    test_size=int(1e3),
    random_state=0,
)
clf = KNeighborsClassifier(n_neighbors=10, n_jobs=1).fit(X_train, y_train)
tic = perf_counter()
y_pred = clf.predict(X_test)
toc = perf_counter()
assert y_pred.shape == y_test.shape
accuracy = np.mean(y_pred == y_test)
print(
    f"n_samples_train: {X_train.shape[0]}, n_samples_test: {X_test.shape[0]}, acc:"
    f" {accuracy:.3f}, duration: {toc - tic:.3f} s"
)
```

**Expected behavior**

Get a high enough classification accuracy (80% or so). See below:

**Output/Screenshots**

- Output for the reference scikit-learn version

```
(sklearn-0.24.2) $ python bench_knn.py 
sklearn 0.24.2 at /scratch/ogrisel/miniforge3/envs/sklearn-0.24.2/lib/python3.9/site-packages/sklearn
[{'architecture': 'Sandybridge',
  'filepath': '/scratch/ogrisel/miniforge3/envs/sklearn-0.24.2/lib/libopenblasp-r0.3.17.so',
  'internal_api': 'openblas',
  'num_threads': 40,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.17'},
 {'filepath': '/scratch/ogrisel/miniforge3/envs/sklearn-0.24.2/lib/libgomp.so.1.0.0',
  'internal_api': 'openmp',
  'num_threads': 40,
  'prefix': 'libgomp',
  'user_api': 'openmp',
  'version': None}]
n_samples_train: 1000000, n_samples_test: 1000, acc: 0.863, duration: 19.761 s
```

- Output for the intelex version:

```
(intel) $ python -m sklearnex bench_knn.py 
Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)
sklearn 0.24.2 at /scratch/ogrisel/miniforge3/envs/intel/lib/python3.8/site-packages/sklearn
[{'filepath': '/scratch/ogrisel/miniforge3/envs/intel/lib/libmkl_rt.so.1',
  'internal_api': 'mkl',
  'num_threads': 20,
  'prefix': 'libmkl_rt',
  'threading_layer': 'intel',
  'user_api': 'blas',
  'version': '2021.3-Product'},
 {'filepath': '/scratch/ogrisel/miniforge3/envs/intel/lib/libomp.so',
  'internal_api': 'openmp',
  'num_threads': 40,
  'prefix': 'libomp',
  'user_api': 'openmp',
  'version': None}]
n_samples_train: 1000000, n_samples_test: 1000, acc: 0.000, duration: 1.373 s
```

Note the 0.0% accuracy: the result is invalid. By varying the size of the dataset, one can get different results but the accuracy is always significantly lower than the scikit-learn's reference implementation.

**Environment:**
 - OS: Ubuntu
 - hardware

<details>

```
$ lscpu
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              40
On-line CPU(s) list: 0-39
Thread(s) per core:  2
Core(s) per socket:  10
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               62
Model name:          Intel(R) Xeon(R) CPU E5-2660 v2 @ 2.20GHz
Stepping:            4
CPU MHz:             2824.454
CPU max MHz:         3000.0000
CPU min MHz:         1200.0000
BogoMIPS:            4400.00
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            25600K
NUMA node0 CPU(s):   0-9,20-29
NUMA node1 CPU(s):   10-19,30-39
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm cpuid_fault epb pti fsgsbase smep erms xsaveopt dtherm ida arat pln pts
```

</details>

 - conda envs
 
<details>

- conda env for the intel version (`conda create -n intel -c intel scikit-learn-intelex`)
```
(intel) $ conda list
# packages in environment at /scratch/ogrisel/miniforge3/envs/intel:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                      1_llvm    conda-forge
blas                      2.110                       mkl    conda-forge
blas-devel                3.9.0                    10_mkl    conda-forge
bzip2                     1.0.8                hb9a14ef_8    intel
ca-certificates           2021.5.30            ha878542_0    conda-forge
certifi                   2021.5.30        py38h578d9bd_0    conda-forge
common_cmplr_lib_rt       2021.3.0             intel_3350    intel
common_cmplr_lic_rt       2021.3.0             intel_3350    intel
cython                    0.29.24          py38h709712a_0    conda-forge
daal4py                   2021.3.0         py38_intel_557    intel
dal                       2021.3.0              intel_557    intel
dpcpp-cpp-rt              2021.3.0             intel_3350    intel
dpcpp_cpp_rt              2021.3.0             intel_3350    intel
icc_rt                    2021.3.0             intel_3350    intel
intel-cmplr-lib-rt        2021.3.0             intel_3350    intel
intel-cmplr-lic-rt        2021.3.0             intel_3350    intel
intel-opencl-rt           2021.3.0             intel_3350    intel
intel-openmp              2021.3.0             intel_3350    intel
intelpython               2021.3.0                      7    intel
joblib                    1.0.1            py38h71958a0_0    intel
libblas                   3.9.0                    10_mkl    conda-forge
libcblas                  3.9.0                    10_mkl    conda-forge
libffi                    3.3                          13    intel
libgcc-ng                 11.1.0               hc902ee8_8    conda-forge
libgfortran-ng            11.1.0               h69a702a_8    conda-forge
libgfortran5              11.1.0               h6c583b3_8    conda-forge
libgomp                   11.1.0               hc902ee8_8    conda-forge
liblapack                 3.9.0                    10_mkl    conda-forge
liblapacke                3.9.0                    10_mkl    conda-forge
libopenblas               0.3.17          pthreads_h8fe5266_1    conda-forge
libstdcxx-ng              11.1.0               h56837e0_8    conda-forge
llvm-openmp               12.0.1               h4bd325d_1    conda-forge
mkl                       2021.3.0              intel_520    intel
mkl-devel                 2021.3.0              intel_520    intel
mkl-include               2021.3.0              intel_520    intel
ncurses                   6.2                  h58526e2_4    conda-forge
numpy                     1.21.1           py38h9894fe3_0    conda-forge
opencl_rt                 2021.3.0             intel_3350    intel
openssl                   1.1.1k               h14c3975_1    intel
pip                       21.2.3             pyhd8ed1ab_0    conda-forge
python                    3.8.11               ha27d850_1    intel
python_abi                3.8                      2_cp38    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
scikit-learn              1.0.dev0                  dev_0    <develop>
scikit-learn-intelex      2021.3.0         py38_intel_557    intel
scipy                     1.7.1            py38h56a6a73_0    conda-forge
setuptools                52.0.0           py38h0d5a7d4_0    intel
sqlite                    3.36.0               h9cd32fc_0    conda-forge
tbb                       2021.3.0              intel_511    intel
tcl                       8.6.10                        1    intel
threadpoolctl             2.2.0              pyh8a188c0_0    conda-forge
tk                        8.6.10               h8e2d9d6_3    intel
wheel                     0.36.2           py38ha11c92b_0    intel
xz                        5.2.5                hcc43529_2    intel
zlib                      1.2.11.1             hb8a9d29_3    intel
```

- For the reference conda-forge env:

```
(sklearn-0.24.2) $ conda list
# packages in environment at /scratch/ogrisel/miniforge3/envs/sklearn-0.24.2:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
ca-certificates           2021.5.30            ha878542_0    conda-forge
certifi                   2021.5.30        py39hf3d152e_0    conda-forge
joblib                    1.0.1              pyhd8ed1ab_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
libblas                   3.9.0               10_openblas    conda-forge
libcblas                  3.9.0               10_openblas    conda-forge
libffi                    3.3                  h58526e2_2    conda-forge
libgcc-ng                 11.1.0               hc902ee8_8    conda-forge
libgfortran-ng            11.1.0               h69a702a_8    conda-forge
libgfortran5              11.1.0               h6c583b3_8    conda-forge
libgomp                   11.1.0               hc902ee8_8    conda-forge
liblapack                 3.9.0               10_openblas    conda-forge
libopenblas               0.3.17          pthreads_h8fe5266_1    conda-forge
libstdcxx-ng              11.1.0               h56837e0_8    conda-forge
ncurses                   6.2                  h58526e2_4    conda-forge
numpy                     1.21.1           py39hdbf815f_0    conda-forge
openssl                   1.1.1k               h7f98852_0    conda-forge
pip                       21.2.3             pyhd8ed1ab_0    conda-forge
python                    3.9.6           h49503c6_1_cpython    conda-forge
python_abi                3.9                      2_cp39    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
scikit-learn              0.24.2           py39h4dfa638_1    conda-forge
scipy                     1.7.1            py39hee8e79c_0    conda-forge
setuptools                49.6.0           py39hf3d152e_3    conda-forge
sqlite                    3.36.0               h9cd32fc_0    conda-forge
threadpoolctl             2.2.0              pyh8a188c0_0    conda-forge
tk                        8.6.10               hed695b0_1    conda-forge
tzdata                    2021a                he74cb21_1    conda-forge
wheel                     0.36.2             pyhd3deb0d_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zlib                      1.2.11            h516909a_1010    conda-forge

```

</details>

**Comment by ogrisel** [comment]
This is a big problem because the benchmarks that are published on the homepage and show 1000x speed-ups or more are probably wrong because this implementation currently produce invalid results:

https://intel.github.io/scikit-learn-intelex/

**Comment by Alexsandruss** [comment]
Thank you for reporting this issue. It is connected not to kNN, but to skleanex implementation of train_test_split, which outputs incorrect numbers in train and test subsets if `train_size + test_size != input_size`. I'm investigating reason of this problem.
Shortened reproducer:
```python
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split

X, y = make_classification(n_samples=20, n_features=4, random_state=0)
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    train_size=10,
    test_size=5,
    random_state=0,
)
print(X_train, y_train, X_test, y_test, sep='\n')
```

If only kNN is patched:
```python
from sklearnex import patch_sklearn
patch_sklearn('knn_classifier')
```
accuracy is same as usual sklearn's implementation.

**Comment by ogrisel** [comment]
Thanks for the investigation.

I fixed by benchmark script as follows to workaround the `train_test_split` bug:

```py
import numpy as np
from subprocess import check_output
import threadpoolctl
from pprint import pprint
from time import perf_counter
from pathlib import Path
import sklearn
from sklearn.neighbors import KNeighborsClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split


sklearn_path = Path(sklearn.__file__).parent
sklearn_version = sklearn.__version__
sklearn_git_branch = (
    check_output(
        f'(cd {sklearn_path} && git symbolic-ref --short HEAD) || echo ""',
        shell=True,
    )
    .decode("utf-8")
    .strip()
)
if sklearn_git_branch:
    sklearn_version = f"{sklearn_version} (git: {sklearn_git_branch})"
print(f"sklearn {sklearn_version} at {sklearn_path}")
pprint(threadpoolctl.threadpool_info())

train_size = int(1e5)
test_size = int(1e5)
n_features = 3
X, y = make_classification(
    n_samples=train_size + test_size,
    n_features=n_features,
    n_redundant=0,
    n_informative=n_features,
    random_state=0,
)
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    train_size=train_size,
    test_size=test_size,
    random_state=0,
)
clf = KNeighborsClassifier(n_neighbors=10, n_jobs=1).fit(X_train, y_train)
tic = perf_counter()
y_pred = clf.predict(X_test)
toc = perf_counter()
assert y_pred.shape == y_test.shape
accuracy = np.mean(y_pred == y_test)
print(
    f"n_samples_train: {X_train.shape[0]}, n_samples_test: {X_test.shape[0]}, acc:"
    f" {accuracy:.3f}, duration: {toc - tic:.3f} s"
)
```

Now DAL/sklearn-intelex indeed returns the same accuracy as scikit-learn and is 10 to 200x faster than sklearn 0.24.2 (depending on the number of samples and features) but not thousands times faster as reported on the homepage. So there is still something I do not understand with those benchmarks.

**Comment by ogrisel** [comment]
Also benchmarking brute force knn with `n_features=3` and kd-tree with `n_features=50` is a bit misleading because one would only use kd-tree in low dimension (e.g. 3) and brute force only in high dim (e.g. 50 or more).

**Comment by ogrisel** [comment]
Closing this issue in favor of the linked issues. Thanks again for the investigations.
